hypothesis-ruby/src/lib.rs

```diff
@@ -1,190 +1,405 @@
+    let repeat_values = HypothesisCoreRepeatValuesStruct::new(
+      min_count.unwrap().to_u64(),
+      max_count.unwrap().to_u64(),
+      expected_count.unwrap().to_f64()
+    );
+
+    Class::from_existing("HypothesisCoreRepeatValues").wrap_data(repeat_values, &*HYPOTHESIS_CORE_REPEAT_VALUES_STRUCT_WRAPPER)
+  }
+
+  fn ruby_hypothesis_core_repeat_values_should_continue(data: AnyObject) -> AnyObject {
+    let mut rdata = data.unwrap();
+    let mut data_source = rdata.get_data_mut(&*HYPOTHESIS_CORE_DATA_SOURCE_STRUCT_WRAPPER);
 
-    def provide(&mut self, data: &mut HypothesisCoreDataSource) -> Option<u64>{
-      data.source.as_mut().and_then(|ref mut source| {
-        distributions::bounded_int(source, self.max_value).ok()
-      })
+    let should_continue = itself
+      .get_data_mut(&*HYPOTHESIS_CORE_REPEAT_VALUES_STRUCT_WRAPPER)
+      ._should_continue(data_source);
+
+    match should_continue {
+      Some(b) => Boolean::new(b).into(),
+      None => NilClass::new().into()
     }
   }
+
+  fn ruby_hypothesis_core_repeat_values_reject() -> NilClass {
+    let repeat_values = itself.get_data_mut(&*HYPOTHESIS_CORE_REPEAT_VALUES_STRUCT_WRAPPER);
+
+    repeat_values.reject();
+
+    NilClass::new()
+  }
+);
+
+pub struct HypothesisCoreBoundedIntegersStruct {
+  max_value: u64
+}
+
+impl HypothesisCoreBoundedIntegersStruct {
+  fn provide(&mut self, data: &mut HypothesisCoreDataSourceStruct) -> Option<u64>{
+    data.source.as_mut().and_then(|ref mut source| {
+      distributions::bounded_int(source, self.max_value).ok()
+    })
+  }
 }
 
-fn mark_child_status(engine: &mut Engine, child: &mut HypothesisCoreDataSource, status: Status) {
-    let mut replacement = None;
-    mem::swap(&mut replacement, &mut child.source);
+wrappable_struct!(HypothesisCoreBoundedIntegersStruct, HypothesisCoreBoundedIntegersStructWrapper, HYPOTHESIS_CORE_BOUNDED_INTEGERS_STRUCT_WRAPPER);
+
+class!(HypothesisCoreBoundedIntegers);
+
+methods!(
+  HypothesisCoreBoundedIntegers,
+  itself,
+
+  fn ruby_hypothesis_core_bounded_integers_new(max_value: Integer) -> AnyObject {
+    let bounded_integers = HypothesisCoreBoundedIntegersStruct { max_value: max_value.unwrap().to_u64() };
+    
+    Class::from_existing("HypothesisCoreBoundedIntegers").wrap_data(bounded_integers, &*HYPOTHESIS_CORE_BOUNDED_INTEGERS_STRUCT_WRAPPER)
+  }
+
+  fn ruby_hypothesis_core_bounded_integers_provide(data: AnyObject) -> AnyObject {
+    let mut rdata = data.unwrap();
+    let data_source = rdata.get_data_mut(&*HYPOTHESIS_CORE_DATA_SOURCE_STRUCT_WRAPPER);
+    let bounded_integers = itself.get_data_mut(&*HYPOTHESIS_CORE_BOUNDED_INTEGERS_STRUCT_WRAPPER);
 
-    match replacement {
-        Some(source) => engine.mark_finished(source, status),
-        None => (),
+    match bounded_integers.provide(data_source) {
+      Some(i) => Integer::new(i as i64).into(),
```

Comment:
`Integer` implements `From<u64>`, what do you think about using `Integer::from(i).into()` instead?